core/src/test/java/org/apache/iceberg/TestPartitionSpecUpdate.java

```diff
@@ -0,0 +1,255 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class TestPartitionSpecUpdate extends TableTestBase {
+
+  private int[] expectedFieldIds;
+
+  private String[] expectedSpecs;
+
+  @Parameterized.Parameters
+  public static Object[][] parameters() {
+    return new Object[][] {
+        new Object[] { 1, new int[]{ 1000, 1001, 1001, 1000, 1000 },
+            new String[]{"  1000: data_bucket_removed: void(2)\n", "  1002: id_bucket_removed: void(1)\n"} },
+        new Object[] { 2, new int[]{ 1001, 1000, 1001, 1002, 1002 }, new String[]{"", ""} },
+    };
+  }
+
+  public TestPartitionSpecUpdate(int formatVersion, int[] expectedFieldIds, String[] expectedSpecs) {
+    super(formatVersion);
+    this.expectedFieldIds = expectedFieldIds;
+    this.expectedSpecs = expectedSpecs;
+  }
+
+  @Test
+  public void testCommitUpdatedSpec() {
+    Assert.assertEquals("[\n" +
+        "  1000: data_bucket: bucket[16](2)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(1000, table.spec().lastAssignedFieldId());
+
+    table.updateSpec().clear()
+        .addBucketField("id", 8)
+        .addBucketField("data", 16)
+        .commit();
```

Comment:
I'm not sure this change makes sense. Why would someone drop the `bucket_16(data)` field only to add it back? I think this should be detected as an error because the field is compatible with a previous field.

If we don't detect this case and throw an exception, then it would be easy for a user to create two fields with the exact same configuration, leading to needless and confusing duplication in the metadata tables.

Reply:
Thanks for the comment. This is true for V1 schema, where we still keep the removed field in the spec. For V2, it won't cause duplication as the pending new field will be exactly the same as the pending removed field (e.g. same field ID) and then equivalent to keep the original untouched.

One case is that if there are a lot of updates for the spec, users might choose to clear it and then add new partition fields. But this seems to bring more troubles than the convenience it offers. So I am thinking we should just remove `clear()` method. If users really want to clear all, they can remove all fields one by one. Any comments on it?

I will add a validation check for the pending changes if users remove a field and then add it back, or add a field and then remove it.